field_encrypt.module

```diff
@@ -21,22 +23,120 @@ function field_encrypt_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $
+          $default_provider = array_shift($valid_provider_ids);
+          $form['field_encrypt']['provider_id'] = [
+            '#type' => 'value',
+            '#value' => $default_provider,
+          ];
+        }
 
-      // Add our encrypted field to the form.
-      $form['encrypt'] = [
-        '#type' => 'checkbox',
-        '#title' => t('Encrypted'),
-        '#description' => t('Makes the field storage encrypted.'),
-        '#default_value' => $field->getThirdPartySetting('field_encrypt', 'encrypt', FALSE),
-      ];
+        // Add settings form for the provider.
+        $form['field_encrypt']['provider_settings'] = array(
+          '#type' => 'container',
+          '#title' => t('Settings'),
+          '#title_display' => FALSE,
+          '#tree' => TRUE,
+          '#prefix' => '<div id="provider-settings">',
+          '#suffix' => '</div>',
+        );
+
+        $selected_provider_id = NULL;
+        $user_input = $form_state->getUserInput();
+        if ($user_input && isset($user_input['field_encrypt']['provider_id'])) {
+          $selected_provider_id = $user_input['field_encrypt']['provider_id'];
+        }
+        if (!$selected_provider_id) {
+          $selected_provider_id = $default_provider;
+        }
+        if ($selected_provider_id) {
+          $provider_settings = $field->getThirdPartySetting('field_encrypt', 'provider_settings', []);
+          $field_encrypt_provider = $provider_manager->createInstance($selected_provider_id, $provider_settings);
+          $plugin_form_state = field_encrypt_create_plugin_form_state($form_state);
+          $form['field_encrypt']['provider_settings'] += $field_encrypt_provider->buildConfigurationForm([], $plugin_form_state);
+          $form_state->setValue(array('field_encrypt', 'provider_settings'), $form_state ->getValues());
+
+          // We add a function to process the form when it is saved.
+          $form['#entity_builders'][] = 'field_encrypt_form_field_add_form_builder';
+        }
+      }
 
-      // We add a function to process the form when it is saved.
-      $form['#entity_builders'][] = 'field_encrypt_form_field_add_form_builder';
     }
   }
 }
 
 /**
+ * AJAX callback to update the dynamic settings on the form.
+ *
+ * @param array $form
+ *   The form definition array for the encryption profile form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current state of the form.
+ *
+ * @return array
+ *   The element to update in the form.
+ */
+function field_encrypt_ajax_update_settings(array &$form, FormStateInterface $form_state) {
+  return $form['field_encrypt']['provider_settings'];
+}
+
+/**
+ * Creates a FormStateInterface object for a plugin.
+ *
+ * @param FormStateInterface $form_state
+ *   The form state to copy values from.
+ *
+ * @return FormStateInterface
+ *   A clone of the form state object with values from the plugin.
+ */
+function field_encrypt_create_plugin_form_state(FormStateInterface $form_state) {
```

Comment:
Is this the bug in which the form doesn't properly clear out when switching between providers? Kind of like what we saw in Encrypt and Key?